File: Test cases/standardLibrary.kt
//
// Copyright 2018 Vinicius Jorge Vendramini
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
typealias PrintContents = Any?

internal fun printTest(contents: PrintContents, testName: String) {
	val firstColumnSize: Int = 40
	val contentsString: String = "${contents}"

	print(contentsString)

	if (contentsString.length < firstColumnSize) {
		for (_0 in contentsString.length until firstColumnSize) {
			print(" ")
		}
	}

	println("(${testName})")
}

fun g(a: Int) {
	printTest(a, "User template")
}

fun main(args: Array<String>) {
	println("Hello, world!")
	println(42)

	val message: String = "A message in a variable."

	println(message)

	val number: Int = 13

	println(number)
	println("Here's a bool literal: ${true}.\nAnd here's a number: ${17}.")
	println("The stored message is: ${message}\nAnd the stored number is: ${number}.")
	print(0)
	println(" (Print)")

	printTest(Math.sqrt(9.0), "Sqrt")

	val string: String = "abcde"
	val bIndex: Int = 1
	val cIndex: Int = 2
	val dIndex: Int = 3
	var variableIndex: Int = cIndex
	val substring: String = "abcd"
	val range: IntRange = IntRange(0, string.length)
	var variableString: String = "abcde"
	val character: Char = 'i'

	printTest(0.toString(), "String(_anyType)")
	printTest("bla".toString(), "String description")
	printTest("".isEmpty(), "String isEmpty")
	printTest("a".isEmpty(), "String isEmpty")
	printTest("".length, "String count")
	printTest("a".length, "String count")
	printTest("abc".firstOrNull()!!, "String first")
	printTest("".firstOrNull(), "String first")
	printTest("abc".lastOrNull()!!, "String last")
	printTest("".lastOrNull(), "String last")
	printTest("0".toDouble(), "String double")
	printTest("1".toDouble(), "String double")
	printTest("0".toFloat(), "String float")
	printTest("1".toFloat(), "String float")
	printTest("0".toULong(), "String uint64")
	printTest("1".toULong(), "String uint64")
	printTest("0".toLong(), "String int64")
	printTest("1".toLong(), "String int64")
	printTest("0".toIntOrNull(), "String int")
	printTest("1".toIntOrNull(), "String int")
	printTest("abcde".dropLast(1), "String dropLast()")
	printTest("abcde".dropLast(2), "String dorpLast(int)")
	printTest("abcde".drop(1), "String dropFirst")
	printTest("abcde".drop(2), "String dropFirst(int)")

	for (index in string.indices) {
		printTest(string[index], "String indices")
	}

	printTest("abcde".substring(0, 4), "String prefix")
	printTest("abcde".substring(0, cIndex), "String prefix(upTo:)")
	printTest("abcde".substring(cIndex), "String index...")
	printTest("abcde".substring(0, cIndex), "String ..<index")
	printTest("abcde".substring(0, cIndex + 1), "String ...index")
	printTest("abcde".substring(bIndex, dIndex), "String index..<index")
	printTest("abcde".substring(bIndex, dIndex + 1), "String index...index")
	printTest(substring, "String String(substring)")
	printTest(string.substring(0, string.length), "String endIndex")
	printTest(string[0], "String startIndex")

	variableIndex -= 1

	printTest(string[variableIndex], "String formIndex(brefore:)")
	printTest(string[cIndex + 1], "String index after")
	printTest(string[cIndex - 1], "String index before")
	printTest(string[cIndex + 2], "String index offset by")
	printTest(substring[bIndex + 1], "String substring index offset by")
	printTest("aaBaBAa".replace("a", "A"), "String replacing occurrences")
	printTest(string.takeWhile { it != 'c' }, "String prefix while")
	printTest(string.startsWith("abc"), "String hasPrefix")
	printTest(string.startsWith("d"), "String hasPrefix")
	printTest(string.endsWith("cde"), "String hasSuffix")
	printTest(string.endsWith("a"), "String hasSuffix")
	printTest(range.start == 0, "String range lowerBound")
	printTest(range.start == string.length, "String range lowerBound")
	printTest(range.endInclusive == 0, "String range upperBound")
	printTest(range.endInclusive == string.length, "String range upperBound")

	val newRange: IntRange = IntRange(0, string.length)

	printTest(newRange.start == 0, "String range uncheckedBounds")
	printTest(newRange.start == string.length, "String range uncheckedBounds")
	printTest(newRange.endInclusive == 0, "String range uncheckedBounds")
	printTest(newRange.endInclusive == string.length, "String range uncheckedBounds")

	variableString += "fgh"

	printTest(variableString, "String append")

	variableString += character

	printTest(variableString, "String append character")
	printTest(string.capitalize(), "String capitalized")
	printTest(string.toUpperCase(), "String uppercased")
	printTest(character.toUpperCase(), "Character uppercased")

	val array2: List<Int> = listOf(2, 1)
	val array4: List<Int> = listOf(2, 1)
	val emptyArray: List<Int> = listOf()
	val stringArray: List<String> = listOf("1", "2", "3")

	val array: MutableList<Int> = mutableListOf(1, 2, 3)
	val array3: MutableList<Int> = mutableListOf(1)
	val arrayOfOptionals: MutableList<Int?> = mutableListOf(1)

	printTest(array, "Array append")

	array.add(4)

	printTest(array, "Array append")

	array.add(0, 0)

	printTest(array, "Array insert")

	array.add(5, 5)

	printTest(array, "Array insert")

	array.add(3, 10)

	printTest(array, "Array insert")

	arrayOfOptionals.add(null)

	printTest(arrayOfOptionals, "Array append nil")

	array3.addAll(array2)

	printTest(array3, "Array append(contentsOf:) constant")

	array3.addAll(array4)

	printTest(array3, "Array append(contentsOf:) variable")
	printTest(emptyArray.isEmpty(), "Array isEmpty")
	printTest(array.isEmpty(), "Array isEmpty")
	printTest(stringArray.joinToString(separator = " => "), "Array joined separator")
	printTest(stringArray.joinToString(separator = ""), "Array joined")
	printTest(array.size, "Array count")
	printTest(stringArray.size, "Array count")

	for (index in array.indices) {
		printTest(array[index], "Array indices")
	}

	printTest(array[0], "Array startIndex")
	printTest(array.size == array.size, "Array endIndex")
	printTest(2 + 1, "Array index after")
	printTest(2 - 1, "Array index before")
	printTest(array.firstOrNull(), "Array first")
	printTest(emptyArray.firstOrNull(), "Array first")
	printTest(array.find { it > 3 }, "Array first where")
	printTest(array.findLast { it > 3 }, "Array last where")
	printTest(array.lastOrNull(), "Array last")

	array.removeAt(0)

	printTest(array, "Array remove first")
	printTest(array.drop(1), "Array drop first")
	printTest(array.dropLast(1), "Array drop last")
	printTest(array.map { it + 1 }.toMutableList(), "Array map")
	printTest(array.flatMap { listOf(it + 1, it + 2) }.toMutableList(), "Array flat map")
	printTest(
		array.map { if (it == 10) { it } else { null } }.filterNotNull().toMutableList(),
		"Array compact map")
	printTest(array.filter { it < 10 }.toMutableList(), "Array filter")
	printTest(array.fold(1) { acc, el -> acc * el }, "Array reduce")

	for ((element1, element2) in array.zip(array2)) {
		printTest(element1, "Array zip")
		printTest(element2, "Array zip")
	}

	printTest(array.indexOfFirst { it > 2 }, "Array firstIndex")
	printTest((array.find { it > 2 } != null), "Array contains where")
	printTest((array.find { it > 2000 } != null), "Array contains where")
	printTest(array.sorted(), "Array sorted")
	printTest(array.contains(10), "Array contains")
	printTest(array.contains(10000), "Array contains")
	printTest(array.indexOf(10), "Array firstIndex of")

	val dictionary: Map<Int, Int> = mapOf(1 to 1, 2 to 2)
	val emptyDictionary: Map<Int, Int> = mapOf()

	printTest(dictionary.size, "Dictionary count")
	printTest(dictionary.isEmpty(), "Dictionary isEmpty")
	printTest(emptyDictionary.isEmpty(), "Dictionary isEmpty")

	val mappedDictionary: List<Int> = dictionary.map { it.value + it.key }

	printTest(mappedDictionary[0], "Dictionary map")
	printTest(mappedDictionary[1], "Dictionary map")
	printTest(Int.MAX_VALUE, "Int max")
	printTest(Int.MIN_VALUE, "Int min")
	printTest(Math.min(0, 1), "Int min(a, b)")
	printTest(Math.min(15, -30), "Int min(a, b)")
	printTest(0..3, "Int ...")
	printTest(-1 until 3, "Int ..<")
	printTest((1.0).rangeTo(3.0), "Double ...")
	printTest(Int.MIN_VALUE until 0, "Recursive matches")

	g(a = 10)
}

File: Test cases/standardLibrary.swift
//
// Copyright 2018 Vinicius Jorge Vendramini
//
// Licensed under the Hippocratic License, Version 2.1;
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// https://firstdonoharm.dev/version/2/1/license.md
//
// To the full extent allowed by law, this software comes "AS IS,"
// WITHOUT ANY WARRANTY, EXPRESS OR IMPLIED, and licensor and any other
// contributor shall not be liable to anyone for any damages or other
// liability arising from, out of, or in connection with the sotfware
// or this license, under any kind of legal claim.
// See the License for the specific language governing permissions and
// limitations under the License.
//

// gryphon output: Test cases/Bootstrap Outputs/standardLibrary.swiftAST
// gryphon output: Test cases/Bootstrap Outputs/standardLibrary.gryphonASTRaw
// gryphon output: Test cases/Bootstrap Outputs/standardLibrary.gryphonAST
// gryphon output: Test cases/Bootstrap Outputs/standardLibrary.kt

import Foundation

// MARK: - Template class declarations
internal class GRYTemplate { // gryphon ignore
	static func dot(_ left: GRYTemplate, _ right: String) -> GRYDotTemplate {
		return GRYDotTemplate(left, right)
	}

	static func dot(_ left: String, _ right: String) -> GRYDotTemplate {
		return GRYDotTemplate(GRYLiteralTemplate(string: left), right)
	}

	static func call(
		_ function: GRYTemplate,
		_ parameters: [GRYParameterTemplate])
		-> GRYCallTemplate
	{
		return GRYCallTemplate(function, parameters)
	}

	static func call(
		_ function: String,
		_ parameters: [GRYParameterTemplate])
		-> GRYCallTemplate
	{
		return GRYCallTemplate(function, parameters)
	}
}

internal class GRYDotTemplate: GRYTemplate { // gryphon ignore
	let left: GRYTemplate
	let right: String

	init(_ left: GRYTemplate, _ right: String) {
		self.left = left
		self.right = right
	}
}

internal class GRYCallTemplate: GRYTemplate { // gryphon ignore
	let function: GRYTemplate
	let parameters: [GRYParameterTemplate]

	init(_ function: GRYTemplate, _ parameters: [GRYParameterTemplate]) {
		self.function = function
		self.parameters = parameters
	}

	//
	init(_ function: String, _ parameters: [GRYParameterTemplate]) {
		self.function = GRYLiteralTemplate(string: function)
		self.parameters = parameters
	}
}

internal class GRYParameterTemplate: ExpressibleByStringLiteral { // gryphon ignore
	let label: String?
	let template: GRYTemplate

	internal init(_ label: String?, _ template: GRYTemplate) {
		if let existingLabel = label {
			if existingLabel == "_" || existingLabel == "" {
				self.label = nil
			}
			else {
				self.label = label
			}
		}
		else {
			self.label = label
		}

		self.template = template
	}

	required init(stringLiteral: String) {
		self.label = nil
		self.template = GRYLiteralTemplate(string: stringLiteral)
	}

	static func labeledParameter(
		_ label: String?,
		_ template: GRYTemplate)
		-> GRYParameterTemplate
	{
		return GRYParameterTemplate(label, template)
	}

	static func labeledParameter(
		_ label: String?,
		_ template: String)
		-> GRYParameterTemplate
	{
		return GRYParameterTemplate(label, GRYLiteralTemplate(string: template))
	}

	static func dot(
		_ left: GRYTemplate,
		_ right: String)
		-> GRYParameterTemplate
	{
		return GRYParameterTemplate(nil, GRYDotTemplate(left, right))
	}

	static func dot(
		_ left: String,
		_ right: String)
		-> GRYParameterTemplate
	{
		return GRYParameterTemplate(nil, GRYDotTemplate(GRYLiteralTemplate(string: left), right))
	}

	static func call(
		_ function: GRYTemplate,
		_ parameters: [GRYParameterTemplate])
		-> GRYParameterTemplate
	{
		return GRYParameterTemplate(nil, GRYCallTemplate(function, parameters))
	}

	static func call(
		_ function: String,
		_ parameters: [GRYParameterTemplate])
		-> GRYParameterTemplate
	{
		return GRYParameterTemplate(nil, GRYCallTemplate(function, parameters))
	}
}

internal class GRYLiteralTemplate: GRYTemplate { // gryphon ignore
	let string: String

	init(string: String) {
		self.string = string
	}
}

internal class GRYConcatenatedTemplate: GRYTemplate { // gryphon ignore
	let left: GRYTemplate
	let right: GRYTemplate

	init(left: GRYTemplate, right: GRYTemplate) {
		self.left = left
		self.right = right
	}
}

internal func + ( // gryphon ignore
	left: GRYTemplate,
	right: GRYTemplate)
	-> GRYConcatenatedTemplate
{
	GRYConcatenatedTemplate(left: left, right: right)
}

internal func + (left: String, right: GRYTemplate) -> GRYConcatenatedTemplate { // gryphon ignore
	GRYConcatenatedTemplate(left: GRYLiteralTemplate(string: left), right: right)
}

internal func + (left: GRYTemplate, right: String) -> GRYConcatenatedTemplate { // gryphon ignore
	GRYConcatenatedTemplate(left: left, right: GRYLiteralTemplate(string: right))
}

// MARK: - Tests

typealias PrintContents = Any // gryphon ignore
// gryphon insert: typealias PrintContents = Any?

func printTest(_ contents: PrintContents, _ testName: String) {
	let firstColumnSize = 40

	let contentsString = "\(contents)"
	print(contentsString, terminator: "")
	if contentsString.count < firstColumnSize {
		for _ in contentsString.count..<firstColumnSize {
			print(" ", terminator: "")
		}
	}
	print("(\(testName))")
}

// Print
print("Hello, world!")
print(42)

let message = "A message in a variable."
print(message)

let number = 13
print(number)

print("Here's a bool literal: \(true).\nAnd here's a number: \(17).")
print("The stored message is: \(message)\nAnd the stored number is: \(number).")

print(0, terminator: "")
print(" (Print)")

// Darwin
printTest(sqrt(9), "Sqrt")

// Numerics
let int: Int = 1
let float: Float = 0.1
let float1: Float = 0.5
let float2: Float = 0.9
let double: Double = 0.1
let double1: Double = 0.5
let double2: Double = 0.9

printTest(Double(int), "Int to Double")
printTest(Float(int), "Int to Float")
printTest(Double(float), "Float to Double")
printTest(Int(float), "Float (0.1) to Int")
printTest(Int(float1), "Float (0.5) to Int")
printTest(Int(float2), "Float (0.9) to Int")
printTest(Float(double), "Double to Float")
printTest(Int(double), "Double (0.1) to Int")
printTest(Int(double1), "Double (0.5) to Int")
printTest(Int(double2), "Double (0.9) to Int")

// String
let string = "abcde"
let bIndex = string.index(string.startIndex, offsetBy: 1) // gryphon value: 1
let cIndex = string.index(string.startIndex, offsetBy: 2) // gryphon value: 2
let dIndex = string.index(string.startIndex, offsetBy: 3) // gryphon value: 3
var variableIndex = cIndex
let substring = "abcde".dropLast() // gryphon value: "abcd"
let range = string.startIndex..<string.endIndex // gryphon value: IntRange(0, string.length)
var variableString = "abcde"
let character: Character = "i"

printTest(String(0), "String(_anyType)")

printTest("bla".description, "String description")

printTest("".isEmpty, "String isEmpty")
printTest("a".isEmpty, "String isEmpty")

printTest("".count, "String count")
printTest("a".count, "String count")

printTest("abc".first!, "String first")
printTest("".first, "String first")

printTest("abc".last!, "String last")
printTest("".last, "String last")

printTest(Double("0"), "String double")
printTest(Double("1"), "String double")

printTest(Float("0"), "String float")
printTest(Float("1"), "String float")

printTest(UInt64("0"), "String uint64")
printTest(UInt64("1"), "String uint64")

printTest(Int64("0"), "String int64")
printTest(Int64("1"), "String int64")

printTest(Int("0"), "String int")
printTest(Int("1"), "String int")

printTest("abcde".dropLast(), "String dropLast()")

printTest("abcde".dropLast(2), "String dorpLast(int)")

printTest("abcde".dropFirst(), "String dropFirst")

printTest("abcde".dropFirst(2), "String dropFirst(int)")

for index in string.indices {
	printTest(string[index], "String indices")
}

printTest("abcde".prefix(4), "String prefix")

printTest("abcde".prefix(upTo: cIndex), "String prefix(upTo:)")

printTest("abcde"[cIndex...], "String index...")

printTest("abcde"[..<cIndex], "String ..<index")

printTest("abcde"[...cIndex], "String ...index")

printTest("abcde"[bIndex..<dIndex], "String index..<index")

printTest("abcde"[bIndex...dIndex], "String index...index")

printTest(String(substring), "String String(substring)")

printTest(string.prefix(upTo: string.endIndex), "String endIndex")

printTest(string[string.startIndex], "String startIndex")

string.formIndex(before: &variableIndex)
printTest(string[variableIndex], "String formIndex(brefore:)")

printTest(string[string.index(after: cIndex)], "String index after")

printTest(string[string.index(before: cIndex)], "String index before")

printTest(string[string.index(cIndex, offsetBy: 2)], "String index offset by")

printTest(substring[substring.index(bIndex, offsetBy: 1)], "String substring index offset by")

printTest("aaBaBAa".replacingOccurrences(of: "a", with: "A"), "String replacing occurrences")

printTest(string.prefix(while: { $0 != "c" }), "String prefix while")

printTest(string.hasPrefix("abc"), "String hasPrefix")
printTest(string.hasPrefix("d"), "String hasPrefix")

printTest(string.hasSuffix("cde"), "String hasSuffix")
printTest(string.hasSuffix("a"), "String hasSuffix")

printTest(range.lowerBound == string.startIndex, "String range lowerBound")
printTest(range.lowerBound == string.endIndex, "String range lowerBound")

printTest(range.upperBound == string.startIndex, "String range upperBound")
printTest(range.upperBound == string.endIndex, "String range upperBound")

let newRange =
	Range<String.Index>(uncheckedBounds:
		(lower: string.startIndex,
		 upper: string.endIndex))
printTest(newRange.lowerBound == string.startIndex, "String range uncheckedBounds")
printTest(newRange.lowerBound == string.endIndex, "String range uncheckedBounds")
printTest(newRange.upperBound == string.startIndex, "String range uncheckedBounds")
printTest(newRange.upperBound == string.endIndex, "String range uncheckedBounds")

variableString.append("fgh")
printTest(variableString, "String append")

variableString.append(character)
printTest(variableString, "String append character")

printTest(string.capitalized, "String capitalized")

printTest(string.uppercased(), "String uppercased")

// Character
printTest(character.uppercased(), "Character uppercased")

// Array
var array = [1, 2, 3] // gryphon ignore
let array2 = [2, 1]
var array3 = [1] // gryphon ignore
let array4 = [2, 1]
var arrayOfOptionals: [Int?] = [1] // gryphon ignore
let emptyArray: [Int] = []
let stringArray = ["1", "2", "3"]
// gryphon insertInMain: val array: MutableList<Int> = mutableListOf(1, 2, 3)
// gryphon insertInMain: val array3: MutableList<Int> = mutableListOf(1)
// gryphon insertInMain: val arrayOfOptionals: MutableList<Int?> = mutableListOf(1)

printTest(array, "Array append")
array.append(4)
printTest(array, "Array append")

array.insert(0, at: 0)
printTest(array, "Array insert")
array.insert(5, at: 5)
printTest(array, "Array insert")
array.insert(10, at: 3)
printTest(array, "Array insert")

arrayOfOptionals.append(nil)
printTest(arrayOfOptionals, "Array append nil")

array3.append(contentsOf: array2)
printTest(array3, "Array append(contentsOf:) constant")

array3.append(contentsOf: array4)
printTest(array3, "Array append(contentsOf:) variable")

printTest(emptyArray.isEmpty, "Array isEmpty")
printTest(array.isEmpty, "Array isEmpty")

printTest(stringArray.joined(separator: " => "), "Array joined separator")

printTest(stringArray.joined(), "Array joined")

printTest(array.count, "Array count")
printTest(stringArray.count, "Array count")

for index in array.indices {
	printTest(array[index], "Array indices")
}

printTest(array[array.startIndex], "Array startIndex")

printTest(array.endIndex == array.count, "Array endIndex")

printTest(array.index(after: 2), "Array index after")

printTest(array.index(before: 2), "Array index before")

printTest(array.first, "Array first")
printTest(emptyArray.first, "Array first")

printTest(array.first(where: { $0 > 3 }), "Array first where")

printTest(array.last(where: { $0 > 3 }), "Array last where")

printTest(array.last, "Array last")

array.removeFirst()
printTest(array, "Array remove first")

printTest(array.dropFirst(), "Array drop first")

printTest(array.dropLast(), "Array drop last")

printTest(array.map { $0 + 1 }, "Array map")

printTest(array.flatMap { [$0 + 1, $0 + 2] }, "Array flat map")

printTest(array.compactMap { $0 == 10 ? $0 : nil }, "Array compact map")

printTest(array.filter { $0 < 10 }, "Array filter")

printTest(array.reduce(1) { acc, el in acc * el }, "Array reduce")

for (element1, element2) in zip(array, array2) {
	printTest(element1, "Array zip")
	printTest(element2, "Array zip")
}

printTest(array.firstIndex(where: { $0 > 2 }), "Array firstIndex")

printTest(array.contains(where: { $0 > 2 }), "Array contains where")
printTest(array.contains(where: { $0 > 2000 }), "Array contains where")

printTest(array.sorted(), "Array sorted")

printTest(array.contains(10), "Array contains")
printTest(array.contains(10000), "Array contains")

printTest(array.firstIndex(of: 10), "Array firstIndex of")

array.removeAll()
printTest(array, "Array remove all")

// Dictionaries
let dictionary = [1: 1, 2: 2]
let emptyDictionary: [Int: Int] = [:]

printTest(dictionary.count, "Dictionary count")

printTest(dictionary.isEmpty, "Dictionary isEmpty")
printTest(emptyDictionary.isEmpty, "Dictionary isEmpty")

let mappedDictionary = dictionary.map { $0.value + $0.key }
printTest(mappedDictionary[0], "Dictionary map")
printTest(mappedDictionary[1], "Dictionary map")

// Int
printTest(Int.max, "Int max")

printTest(Int.min, "Int min")

printTest(min(0, 1), "Int min(a, b)")
printTest(min(15, -30), "Int min(a, b)")

printTest(0...3, "Int ...")

printTest(-1..<3, "Int ..<")

// Double
printTest(1.0...3.0, "Double ...")

//
// Recursive matches
printTest(Int.min..<0, "Recursive matches")

////////////////////////////////////////////////////////////////////////////////////////////////////
// MARK: - Test templates details

// Define the templates
func gryphonTemplates() {
	var _int = 0
	var _a = A()

	f(of: _int)
	"g(a = _int)"

	_a.string.first
	GRYTemplate.dot(.dot("_a", "string"), "firstOrNull()")

	_ = G.i
	_ = GRYTemplate.dot("G", "I")

	_ = G.H.k
	_ = GRYTemplate.dot(.dot("G", "H"), "K")
}

//////////////////////////////////////////////////////////
// Auxiliary declarations for the following tests
class A {
	let string: String = ""
}

class B {
	var description: String = ""
}

class C: CustomStringConvertible {
	var description: String = ""
}

func f(of a: Int) { // gryphon ignore
	printTest(a, "User template")
}

// gryphon insert: fun g(a: Int) {
// gryphon insert: 	printTest(a, "User template")
// gryphon insert: }

enum G { // gryphon ignore
	case i

	enum H {
		case k
	}
}

// gryphon insert:
// gryphon insert: enum class G {
// gryphon insert: 	I;
// gryphon insert:
// gryphon insert: 	enum class H {
// gryphon insert: 		K;
// gryphon insert: 	}
// gryphon insert: }

//////////////////////////////////////////////////////////
// User defined templates
f(of: 10)

// Adding `?` in dot chains declared by templates
let maybeA: A? = nil
if let a = maybeA, let b = a.string.first {
	// ...
}

// Adding `@map` labels in closures called by functions in templates
print([1, 2, 3].map { (a: Int) -> Int in
	if a > 1 {
		return 1
	}
	else {
		return 2
	}
})

// Adding labels for nested closures
print([1, 2, 3].map { (a: Int) -> [Int] in
	if a > 1 {
		return [1, 2, 3].filter { (b: Int) -> Bool in
			if b > 1 {
				return true
			}
			else {
				return false
			}
		}
	}
	else {
		return [2]
	}
})

//////////////////////////////////////////////////////////
// Inheritance checking

// B doesn't conform to CustomStringConvertible, so this is `B().description`
let description1 = B().description

// C conforms to CustomStringConvertible, so this is `C().toString()`
let description2 = C().description

// Local variables don't get translated
let description: String = ""

// Nested classes get translated
class D {
	class E: CustomStringConvertible {
		var description: String = ""
	}
}

//////////////////////////////////////////////////////////
// Check implicit TypeExpressions match explicit DotExpressions
// (if they don't match this won't be capitalized)
let w: G = .i
let y: G.H = .k
